_posts/2022-05-12-data-pipelines-in-Databricks.markdown

```diff
@@ -0,0 +1,242 @@
+
+I was recently working for a client that required a data pipeline in Databricks that would conform to the DevOps way of working. This pipeline would need to have an RBAC model to conform to least privilege, monitoring and observability and include testing of transformed data.
+How did we achieve this? In this blog post I will bring you along on our journey of how we were able to implement this data pipeline.
+
+<br />
+
+### Delta Lake
+
+A Delta lake enables developers to create Lakehouse architectures on top of data lakes and is open source. Delta lakehouse combines elements from a data warehouse and a data lake.
+
+![data lakehouse](/img/blog/databricks-data-pipeline/data-lakehouse.png)
+
+### Databricks
+
+Databricks provides a solution for ingestion, processing and scheduling allowing developers to run ETL workloads directly on a data lake.
+
+An important part of creating a data solution in Databricks is setting up your data pipeline. A data pipeline in Databricks consist of one or multiple transformation tasks where you transform your data through Raw (Bronze), Cleansed (Silver) and Curated (Gold) layers.
+
+Databricks documentation recommend using Delta Lake patterns for data transformations where you move your data from a Bronze layer to a Gold layer.
+
+- Bronze
+  - Ingest raw data
+- Silver
+  - Deduplicate data
+  - Enforce schemas
+- Gold
+  - Perform business-specific transforms to data
+  - Aggregate data
+
+<br/>
+
+![delta lake](/img/blog/databricks-data-pipeline/delta-lake.png)
+
+<br/>
+
+### Multi task jobs v delta live jobs
+
+<br/>
+
+In Databricks you can use two different methods for creating data pipelines, multi-task jobs or delta live tables. What are the differences and which method did we choose?
+
+Multi-task jobs:
+
+- This is suited for batch jobs. Multi task jobs need to wait for a task to finish before it can move onto the next task. If you have a streaming job the multi task job cannot determine when to finish a task and move onto the next.
+
+- Allows you to execute notebooks in a flow
+
+This allows you to set your bronze - silver layer dependencies and run the pipeline in that set sequence.
+This allows you to have a single workflow or parallel workflows so you can build your data pipeline in a way that suits your individual needs.
+
+Databricks enables you to run your data pipeline on a schedule that is as simple as setting up a cron job in your IaC. The alternative is to go into the Databricks UI and manually trigger your data pipeline.
+
+Delta live tables:
+
+- This is suited for streaming jobs.
+- More suited for creating ETL pipelines.
+
+Although delta live would have been better suited for our needs we did not use delta live as it was still in preview and not ready for production workflows at the time.
+
+## Challenges
+
+Some of the challenges we faced were:
+
+Dags are autogenerated, this was a challenge because we have no flexibility to move the dags so the pipeline can look messy.
```

Comment:
When you create a multi task job in Databricks, dags are auto generated. This was a challenge because as we created more tasks and more tasks dependencies, the pipeline would start looking very messy and dependency arrows were colliding into each other. There is no flexibility to move the dags around or to adjust how they are generated